src/entities/join/weighted/helpers.ts

```diff
@@ -0,0 +1,135 @@
+export function getAmountsIn(input: JoinInput, poolAssets: string[]): bigint[] {
+    return poolAssets.map((asset) => {
+        let tokenIn: TokenAmount | undefined;
+        switch (input.kind) {
+            case JoinKind.Init:
+                tokenIn = input.initAmountsIn.find(
+                    (t) => t.token.wrapped === asset,
+                );
+                break;
+            case JoinKind.Proportional:
+                if (input.refAmountIn.token.wrapped === asset)
+                    tokenIn = input.refAmountIn;
+                // TODO: calculate proportional amounts based on reference token
+                break;
+            case JoinKind.Unbalanced:
+                tokenIn = input.amountsIn.find(
+                    (t) => t.token.wrapped === asset,
+                );
+                break;
+            case JoinKind.SingleAsset:
+                if (input.amountIn.token.wrapped === asset)
+                    tokenIn = input.amountIn;
+                break;
+        }
+        return tokenIn?.amount ?? 0n;
+    });
+}
+
+export function getUserData(input: JoinInput, amountsIn: bigint[]): Address {
+    let userData: Address;
+    switch (input.kind) {
+        case JoinKind.Init: {
+            userData = WeightedEncoder.joinInit(amountsIn);
+            break;
+        }
+        case JoinKind.Proportional:
+        case JoinKind.Unbalanced:
+        case JoinKind.SingleAsset: {
+            userData = WeightedEncoder.joinGivenIn(amountsIn, 0n);
+            break;
+        }
+        case JoinKind.ExactOut: {
+            userData = WeightedEncoder.joinGivenOut(input.bptOut.amount);
+            break;
+        }
+        default:
+            throw new Error('Invalid join kind');
+    }
+    return userData;
+}
+
+export function getAssets(input: JoinInput, poolAssets: Address[]): Address[] {
+    let nativeTokenIn: TokenAmount | undefined;
+    switch (input.kind) {
+        case JoinKind.Init:
+            nativeTokenIn = input.initAmountsIn.find(
+                (t) => t.token.address === ZERO_ADDRESS,
+            );
+            break;
+        case JoinKind.Proportional:
+            if (input.refAmountIn.token.address === ZERO_ADDRESS)
+                nativeTokenIn = input.refAmountIn;
+            break;
+        case JoinKind.Unbalanced:
+            nativeTokenIn = input.amountsIn.find(
+                (t) => t.token.address === ZERO_ADDRESS,
+            );
+            break;
+        case JoinKind.SingleAsset:
+            if (input.amountIn.token.address === ZERO_ADDRESS)
+                nativeTokenIn = input.amountIn;
+            break;
+    }
+    let assets = [...poolAssets];
+
+    if (nativeTokenIn !== undefined) {
+        const wrappedAssetIndex = poolAssets.findIndex(
+            // TODO: this assumes that the user provided wrapped address for input tokens - is that a fair assumption?
```

Comment:
This is another tough one, as it is creating layers of requirements that are very easy to forget or just not even know about.

Reply:
yeah, I think we should not expect the user to provide that info - if we decided to limit the user input to pool tokens only, handling this native/wrapped addresses won't be needed anymore, so we should be ok
if we do accept `zero address` as input, we'll likely need a native/wrapped mapping hardcoded within b-sdk (as we currently have on the SDK)